kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ServiceDiscoveryClient.scala

```diff
@@ -0,0 +1,139 @@
+
+import org.apache.kyuubi.KyuubiException
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.ha.HighAvailabilityConf.HA_ZK_CONN_MAX_RETRIES
+import org.apache.kyuubi.ha.HighAvailabilityConf.HA_ZK_CONN_MAX_RETRY_WAIT
+import org.apache.kyuubi.ha.HighAvailabilityConf.HA_ZK_NAMESPACE
+import org.apache.kyuubi.ha.client.ServiceDiscovery
+import org.apache.kyuubi.ha.client.ServiceDiscovery.createServiceNode
+import org.apache.kyuubi.ha.client.ZooKeeperClientProvider.buildZookeeperClient
+import org.apache.kyuubi.ha.client.zookeeper.ServiceDiscoveryClient.connectionChecker
+import org.apache.kyuubi.util.ThreadUtils
+
+class ServiceDiscoveryClient(serviceDiscovery: ServiceDiscovery) extends Logging {
+
+  /**
+   * a pre-defined namespace used to publish the instance of the associate service
+   */
+  protected var _namespace: String = _
+
+  private lazy val instance: String = serviceDiscovery.fe.connectionUrl
+  private var zkClient: CuratorFramework = _
+  private var serviceNode: PersistentNode = _
+
+  def namespace: String = _namespace
+
+  def createClient(conf: KyuubiConf): Unit = {
+    _namespace = conf.get(HA_ZK_NAMESPACE)
+    val maxSleepTime = conf.get(HA_ZK_CONN_MAX_RETRY_WAIT)
+    val maxRetries = conf.get(HA_ZK_CONN_MAX_RETRIES)
+    zkClient = buildZookeeperClient(conf)
+    zkClient.getConnectionStateListenable.addListener(new ConnectionStateListener {
+      private val isConnected = new AtomicBoolean(false)
+
+      override def stateChanged(client: CuratorFramework, newState: ConnectionState): Unit = {
+        info(s"Zookeeper client connection state changed to: $newState")
+        newState match {
+          case CONNECTED | RECONNECTED => isConnected.set(true)
+          case LOST =>
+            isConnected.set(false)
+            val delay = maxRetries.toLong * maxSleepTime
+            connectionChecker.schedule(
+              new Runnable {
+                override def run(): Unit = if (!isConnected.get()) {
+                  error(s"Zookeeper client connection state changed to: $newState, but failed to" +
+                    s" reconnect in ${delay / 1000} seconds. Give up retry. ")
+                  serviceDiscovery.stopGracefully()
+                }
+              },
+              delay,
+              TimeUnit.MILLISECONDS)
+          case _ =>
+        }
+      }
+    })
+    zkClient.start()
+  }
+
+  def registerService(conf: KyuubiConf): Unit = {
+    serviceNode = createServiceNode(conf, zkClient, namespace, instance)
+    // Set a watch on the serviceNode
+    val watcher = new DeRegisterWatcher
+    if (zkClient.checkExists.usingWatcher(watcher).forPath(serviceNode.getActualPath) == null) {
+      // No node exists, throw exception
+      throw new KyuubiException(s"Unable to create znode for this Kyuubi " +
+        s"instance[${instance}] on ZooKeeper.")
+    }
+  }
+
+  // close the EPHEMERAL_SEQUENTIAL node in zk
```

Comment:
let make some comment more common for this func, and move the current comment inside the func